ext/pdo/pdo_sql_parser.re

```diff
@@ -386,24 +380,23 @@ rewrite:
 
 			/* check if bound parameter is already available */
 			if (!strcmp(name, "?") || (p = zend_hash_str_find_ptr(stmt->bound_param_map, name, plc->len)) == NULL) {
-				spprintf(&idxbuf, 0, tmpl, bind_no++);
+				idxbuf = strpprintf(0, tmpl, bind_no++);
 			} else {
-				idxbuf = estrdup(p);
+				idxbuf = zend_string_init(p, strlen(p), 0);
 				skip_map = 1;
 			}
 
 			plc->quoted = idxbuf;
-			plc->qlen = strlen(plc->quoted);
 			plc->freeq = 1;
-			newbuffer_len += plc->qlen;
+			newbuffer_len += ZSTR_LEN(plc->quoted);
 
 			if (!skip_map && stmt->named_rewrite_template) {
 				/* create a mapping */
-				zend_hash_str_update_mem(stmt->bound_param_map, name, plc->len, idxbuf, plc->qlen + 1);
+				zend_hash_str_update_mem(stmt->bound_param_map, name, plc->len, ZSTR_VAL(plc->quoted), ZSTR_LEN(plc->quoted) + 1);
```

Comment:
It would be more elegant to switch bound_param_map to store zend_string* instead, which means you wouldn't need to create a separate copy of the string to store in this map (_mem performs a copy).

Reply:
Will look into this

Reply:
bound_param_map is a HashTable, so I'm not sure how one can go about to do that (I'm rather unfamiliar with the whole HashTable API and how it works under the hood).

Reply:
In this case, you would modify free_param_name to do a zend_string_release instead and use zend_hash_str_update_ptr instead of zend_hash_str_update_mem to insert the zend_string pointer.